modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/UnsortedDistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,211 @@
+        Collection<ClusterNode> nodes) {
+        super(fut);
+
+        this.reqId = reqId;
+        this.pageRequester = pageRequester;
+
+        synchronized (queueLock()) {
+            for (ClusterNode node : nodes)
+                subgrid.add(node.id());
+        }
+
+        cctx = fut.cctx;
+
+        pageStream = new PageStream();
+
+        this.fut = fut;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean hasNext() throws IgniteCheckedException {
+        return pageStream.hasNext();
+    }
+
+    /** {@inheritDoc} */
+    @Override public R next() throws IgniteCheckedException {
+        return pageStream.next();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addPage(@Nullable UUID nodeId, Collection<R> data) {
+        pageStream.addPage(data);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onLastPage() {
+        super.onLastPage();
+
+        loadAllowed = false;
+
+        firstPageLatch.countDown();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancel() {
+        Collection<ClusterNode> allNodes = cctx.discovery().allNodes();
+        Collection<ClusterNode> nodes;
+
+        synchronized (queueLock()) {
+            nodes = F.retain(allNodes, true,
+                new P1<ClusterNode>() {
+                    @Override public boolean apply(ClusterNode node) {
+                        return !cctx.localNodeId().equals(node.id()) && subgrid.contains(node.id());
+                    }
+                }
+            );
+
+            rcvd.clear();
+            subgrid.clear();
+        }
+
+        pageRequester.cancelQueryRequest(reqId, nodes, fut.fields());
+
+        pageStream.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void loadPage() {
+        assert !Thread.holdsLock(queueLock());
+
+        Collection<UUID> nodes = null;
+
+        synchronized (queueLock()) {
+            // Loads only queue is empty to avoid memory consumption on additional pages.
+            if (!pageStream.queue.isEmpty())
+                return;
+
+            if (loadAllowed && rcvd.containsAll(subgrid)) {
+                rcvd.clear();
+
+                nodes = new ArrayList<>(subgrid);
+            }
+        }
+
+        if (nodes != null)
+            pageRequester.requestPages(reqId, fut, nodes, false);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void loadAll() throws IgniteInterruptedCheckedException {
+        assert !Thread.holdsLock(queueLock());
+
+        U.await(firstPageLatch);
+
+        Collection<UUID> nodes = null;
+
+        synchronized (queueLock()) {
+            if (loadAllowed && !subgrid.isEmpty())
+                nodes = new ArrayList<>(subgrid);
+        }
+
+        if (nodes != null)
+            pageRequester.requestPages(reqId, fut, nodes, true);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onPage(@Nullable UUID nodeId, boolean last) {
```

Comment:
onPage with `boolean last` and onLastPage() looks confusing.
Looks like an architectural lack.
What is the difference between these methods?

Reply:
Merged them to single method.